usb: patches to resolve issues around Kconfig option USB_UART_CONSOLE

jfischer-no opened the pull request:
Patches to resolve issues around Kconfig option USB_UART_CONSOLE.

Alternative to https://github.com/zephyrproject-rtos/zephyr/pull/40192
Alternative to https://github.com/zephyrproject-rtos/zephyr/pull/39987

jfischer-no committed:
usb: cdc_acm: use same init level and priority as serial drivers

Use same init level and priority as serial drivers.
Align priority to the changes in
commit ad1450510aff ("drivers: serial: Refactor drivers
to use shared init priority Kconfig")

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: limit scope of CONFIG_USB_UART_CONSOLE

Do not change init level and priority of console driver
if Kconfig option CONFIG_USB_UART_CONSOLE is enabled because
CDC ACM UART driver now uses the same init level and priority
as regular serial driver.

Replace CONFIG_USB_UART_CONSOLE with
DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_...), zephyr_cdc_acm_uart)
in the source code.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: deprecate Kconfig option USB_UART_CONSOLE

Kconfig option USB_UART_CONSOLE does not have any dependencies
on console drivers, but is used to select backend specific
Kconfig option in subsys/shell/backends/Kconfig.backends.

Move Kconfig option USB_UART_CONSOLE to CDC ACM configuration,
deprecate it and update description.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jakub-uC commented:
@jfischer-no : May you please let me know if you have tested this fix with nrf52840 dongle?

jfischer-no commented on subsys/usb/class/Kconfig.cdc:
```diff
@@ -53,6 +53,12 @@ config CDC_ACM_DTE_RATE_CALLBACK_SUPPORT
+config USB_UART_CONSOLE
+	bool "Use CDC ACM UART as backend for console, shell, or logging"
+	help
+	  This option is only used to enable relevant options in backend
+	  configuration (if any), and for backward compatibility.
```
No, it is in use and is investigated, see commit message:
"Kconfig option USB_UART_CONSOLE does not have any dependencies
on console drivers, but is used to select backend specific
Kconfig option in subsys/shell/backends/Kconfig.backends.
Keep the name of the option USB_UART_CONSOLE but move it
to CDC ACM configuration and update description."

jfischer-no replied:
Commit was reworked. I will remove Kconfig option USB_UART_CONSOLE and dependency in subsys/shell/backends/Kconfig.backends in subsequent PR because MCUboot needs to be adapted as well.

jfischer-no commented on drivers/console/Kconfig:
```diff
@@ -34,7 +34,7 @@ config CONSOLE_HANDLER
 config CONSOLE_INIT_PRIORITY
 	int "Console init priority"
-	default 95 if USB_UART_CONSOLE || UART_MUX
+	default 95 if UART_MUX
```
> So what has changed, since console can now be initialized at `60` instead ?
> This commit only changed the symbol used: [d8bc37b](https://github.com/zephyrproject-rtos/zephyr/commit/d8bc37beeb4d1eb4fdc47f132c15915f73a9ef50)
> but the default value is still `50`
> `SERIAL_INIT_PRIORITY` defaults to `KERNEL_INIT_PRIORITY_DEVICE` default == `50`

CDC ACM UART driver has been reworked a bit and does not block when USB device support is not initialized, (behaves like regular UART).

>Or should the setting be promptless in certain cases ?

I am in favor of this option.

jfischer-no commented on subsys/usb/class/Kconfig.cdc:
```diff
@@ -53,6 +53,12 @@ config CDC_ACM_DTE_RATE_CALLBACK_SUPPORT
 	  by Arduino style programmers to reset the device into the
 	  bootloader.
 
+config USB_UART_CONSOLE
```
No, it is there for reasons.

jfischer-no replied:
deprecated now

jfischer-no commented:
> @jfischer-no : May you please let me know if you have tested this fix with nrf52840 dongle?

yes I am testing it with nrf52840dongle_nrf52840

jfischer-no committed:
usb: add option to initialize USB device support at boot

Possibility to initialize USB device support is useful when
only CDC ACM class is enabled and CDC ACM UART is used as
backend for console, shell, or logging.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
shields: add generic shield to use CDC ACM UART as backend

Add  generic shields that provide devicetree and configuration
overlays, and configure USB device stack so that CDC ACM UART
can be used as backend for console, logging, or shell.
This approach allows us to avoid many identical overlays in
samples and tests directories, and cumbersome board configurations.

These shields use Kconfig option CONFIG_USB_DEVICE_INITIALIZE_AT_BOOT
and is mainly intended to be used with boards like
nrf52840dongle_nrf52840 or bl654_usb.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: shell_module: use generic CDC ACM shield

Remove CDC ACM UART devicetree and config overlay, and explicit
USB support initialization in favor of generic CDC ACM shield.
For shell support via USB CDC ACM, this example is to be built
as follows:

west build -b nrf52840dongle_nrf52840
  samples/subsys/shell/shell_module -- -DSHIELD=cdc_acm_shell

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests: bluetooth: shell: use generic CDC ACM shield

Remove CDC ACM UART devicetree and config overlay, and explicit
USB support initialization in favor of generic CDC ACM shield.
For shell support via USB CDC ACM, this example is to be built
as follows:

west build -b nrf52840dongle_nrf52840
  tests/bluetooth/shell -- -DSHIELD=cdc_acm_shell

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests: uart_basic_api: use generic CDC ACM shield

Remove CDC ACM UART devicetree and config overlay, and explicit
USB support initialization in favor of generic CDC ACM shield.
For console support via USB CDC ACM, this example is to be built
as follows:

west build -b reel_board
  tests/drivers/uart/uart_basic_api -- -DSHIELD=cdc_acm_console

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: sensortile_box: use generic CDC ACM shield

Remove CDC ACM UART devicetree overlay, and explicit
USB support initialization in favor of generic CDC ACM shield.
For console support via USB CDC ACM, this example is to be built
as follows:

west build -b sensortile_box
  samples/boards/sensortile_box -- -DSHIELD=cdc_acm_console

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: console: fix configuration

This sample uses console driver and therefore
Kconfig option CONFIG_CONSOLE and CONFIG_UART_CONSOLE
are required.

Kconfig option CONFIG_USB_UART_CONSOLE no longer has
any influence here and should be removed.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards: clean up CDC ACM configuration for nRF52840 based boards

Remove CDC ACM UART devicetree nodes, chosen properties and explicit
USB support configuration in favor of generic CDC ACM shield.
For console support via USB CDC ACM, in-tree samples for this boards
is to be built as follows:

west build -b nrf52840dongle_nrf52840
  samples/basic/threads -- -DSHIELD=cdc_acm_console

For shell support via USB CDC ACM, in-tree samples for this boards
is to be built as follows:

west build -b bl654_usb
  samples/subsys/shell/shell_module -- -DSHIELD=cdc_acm_shell

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: remove the remains of USB_UART_CONSOLE

Since there are no more users and dependencies of
Kconfig option USB_UART_CONSOLE in the tree, remove the remains.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on tests/bluetooth/shell/src/main.c:
```diff
@@ -122,10 +123,20 @@ static void hrs_notify(void)
-	if (IS_ENABLED(CONFIG_USB_UART_CONSOLE)) {
-		usb_enable(NULL);
-		k_sleep(K_SECONDS(2));
+#if DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_shell_uart), zephyr_cdc_acm_uart)
```
This part is removed now.

jfischer-no committed:
shields: cdc_acm: add support for zephyr,bt-c2h-uart property

Add support for chosen property "zephyr,bt-c2h-uart".

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

erwango commented on tests/drivers/uart/uart_basic_api/testcase.yaml:
```diff
@@ -19,8 +19,7 @@ tests:
   drivers.uart.cdc_acm:
-    extra_args: OVERLAY_CONFIG="overlay-usb.conf"
-                DTC_OVERLAY_FILE="usb.overlay"
+    extra_args: SHIELD="cdc_acm_console"
```
A question here, I like the general direction, but I'd like to know how tests will work with boards that already declare usb for value of `zephyr,console` chosen.

jfischer-no replied:
AFAIK last property definition wins. With this PR, there are no more boards that define chosen property with reference to "zephyr,cdc-acm-uart" compatible node. But it works even if it is the case and shield is used, then there would be two CDC ACM devices, one of them would be used by the test.

carlescufi reviewed (commented):
I am not against this and I have tested it and works fine. but maybe we should rethink the word `SHIELD` here. Perhaps we could come up with something that describes both meta-configurations and actual hardware shields. Some new name, like `METACFG`?

Imagine we used it for other purposes too, like:
```
west build samples/hello_world -- -DMETACFG=usb_cdc_acm -DMETACFG=use_mcuboot -DMETACFG=my_hw_shield
```

tejlmand commented on boards/arm/arduino_nano_33_ble/Kconfig.board:
```diff
@@ -5,16 +5,6 @@ config BOARD_ARDUINO_NANO_33_BLE
 	bool "Arduino Nano 33 BLE board"
 	depends on SOC_NRF52840_QIAA
 
-config BOARD_ARDUINO_NANO_33_BLE_EN_USB_CONSOLE
```
so all boards cleaned up in this commit 9c034972b1e618bfddc7365360ba10bd924e9e54 must now use `-DSHIELD=cdc_acm_console`

That sounds like a step backward in user-friendliness if those boards are expected to always use the usb as console.

For example, this board `bl654_usb` doesn't seem to have any other option than using USB for console:
https://docs.zephyrproject.org/latest/boards/arm/bl654_usb/doc/bl654_usb.html

@kieran-mackey FYI.

jfischer-no replied:
No, it is a step forward, only with these patches it becomes possible to use CDC ACM UART (_not usb_) as backend for console. The configuration of this board in the tree is not usable.

tejlmand replied:
>  The configuration of this board in the tree is not usable.

Which board or all boards ?

Read my comment:
> so all boards cleaned up in this commit 9c03497 must now use -DSHIELD=cdc_acm_console

i'm refering to the entire commit 9c03497, not a single board, but github requires me to put the comment on a single line.
That is, my comment covers all those boards you are modifying in that commit:
- arduino_nano_33_ble
- bl654_usb
- degu_evk
- nrf52840dongle_nrf52840

are you saying that the current configuration in Zephyr of those boards are not working for any of them ?

tejlmand replied:
as far as I'm aware, the general problem started with this cleanup:
https://github.com/zephyrproject-rtos/zephyr/commit/d12331ea4b5d8660f50f1dde007e1441849bc361
which again was fixing:
https://github.com/zephyrproject-rtos/zephyr/issues/38403
caused by this:
https://github.com/zephyrproject-rtos/zephyr/pull/37512

So maybe we should go back to the origin of the problem, instead of completely changing how users must use the boards.

tejlmand reviewed (changes requested):
Main issue I see is the `SHIELD=cdc_acm_console`, please see detailed comments.

The reason for this PR is stated to be
> Patches to resolve issues around Kconfig option USB_UART_CONSOLE.
https://github.com/zephyrproject-rtos/zephyr/pull/40220#issue-776750382

but it is not clear to me why this series of commits are needed to solve the issue.

Is is possible to split this PR into the part that is needed to solve the particular issue, and then a second part with the cdc acm console enhancement ?

If not, then please make a proper description why the big rewrite to the system is needed.

jfischer-no commented:
> If not, then please make a proper description why the big rewrite to the system is needed.

Each commit has a description, please read through it. I see no reason to copy it again in PR description.

jfischer-no commented on boards/shields/cdc_acm/doc/index.rst:
```diff
@@ -0,0 +1,64 @@
+.. _cdc_acm_shield:
+
+Generic shields for CDC ACM UART
```
I am convinced that shield approach is a better fit for CDC ACM UART, similar to how one would connect a real UART controller through I2C/SPI using a shield.

jfischer-no commented on samples/boards/sensortile_box/app.overlay:
```diff
@@ -1,18 +0,0 @@
-/*
```
No, they would look in samples/boards/sensortile_box/README.rst, preferably before, and see how to build it. https://github.com/zephyrproject-rtos/zephyr/pull/40220/commits/461b174e3494d1782c572fd0538a26ff4cc1ca83#diff-9b13c471892c9de2d5bb33dce8bfaf55dd08edb938cb2ed9642f77bb32496000R40

tejlmand replied:
but having to apply an extra argument, in a place where `app.overlay` used to work, seems to be a step backward.

A sample should preferable be able to built in normal mode without extra args.

mbolivar-nordic commented:
> Each commit has a description, please read through it. I see no reason to copy it again in PR description.

A pull request description in GitHub is like the cover letter for a patch series sent via email. It is an overall justification for the work, along with an overview of the individual commits. It clearly has a purpose.

mbolivar-nordic reviewed (changes requested):
I understand that this is probably the easiest way to accomplish the goal of adding some Kconfig and DT overlays into the build using a single variable setting (`-DSHIELD=...`), but this is really a confusing abuse of the system.

I think what's going on here is that a "shield" is a special case of something more general that has what you want (@carlescufi 's "meta-config"), and you are using the existing mechanism because we haven't solved the real problem.

As discussed offline I think this is similar to the problem @JordanYates was trying to solve for mcuboot, which he discussed at the Zephyr Developers Summit (ZDS). We should try to solve the real problem, not just abuse shields, especially since there is also ongoing work to try to solve the multi-image build problem for the special case of mcuboot.

At ZDS, we discussed having an "include path" or a list of search directories for places that would contain both Kconfig and DTS files to apply to the build. Everyone agreed about the general idea, but nobody could agree what the call that thing, and we never agreed on anything. Let's try again; we have enough interested people that I hope we can do it this time.

tejlmand commented:
> > If not, then please make a proper description why the big rewrite to the system is needed.
> 
> Each commit has a description, please read through it. I see no reason to copy it again in PR description.

I did, and for example I see this in one of the commit messages:
> Add  generic shields that provide devicetree and configuration
> overlays, and configure USB device stack so that CDC ACM UART
> can be used as backend for console, logging, or shell.
> This approach allows us to avoid many identical overlays in
> samples and tests directories, and cumbersome board configurations.
>

This talks about what you do, and a little why.
But it doesn't tell me how this relates to the problem this PR is solving, which is:
https://github.com/zephyrproject-rtos/zephyr/pull/40220#issue-1048848671
> Patches to resolve issues around Kconfig option USB_UART_CONSOLE.

To me particular commit message I just posted seems to solve a very different issue imho.

jfischer-no commented:
> I understand that this is probably the easiest way to accomplish the goal of adding some Kconfig and DT overlays into the build using a single variable setting (`-DSHIELD=...`), but this is really a confusing abuse of the system.

I do not abuse anything, I am adding a shield for an emulated UART controller. It is also not an abuse to have emulated boards like qemu_cortex_m0.

> I think what's going on here is that a "shield" is a special case of something more general that has what you want (@carlescufi 's "meta-config"), and you are using the existing mechanism because we haven't solved the real problem.
> 
> As discussed offline I think this is similar to the problem @JordanYates was trying to solve for mcuboot, which he discussed at the Zephyr Developers Summit (ZDS). We should try to solve the real problem, not just abuse shields, especially since there is also ongoing work to try to solve the multi-image build problem for the special case of mcuboot.

@mbolivar-nordic, @tejlmand, @carlescufi 
You are now connecting this with something that is not even defined. Sorry, but this sounds like long discourse with little output, especially when it comes to multi-image or mcuboot. I have a concrete issue and a solution now. The alternative is then to change the boards, without debug probe, to minimal configuration (remove all USB device support configuration option) which is a kind of tacit rule.

I opened #40428 to remove USB_UART_CONSOLE option.
